assets/src/components/form/inputs/checkbox-input.js

```diff
@@ -0,0 +1,115 @@
+/**
+ * External imports
+ */
+import { Component } from 'react';
+import PropTypes from 'prop-types';
+
+/**
+ * Internal imports
+ */
+import { OptionCheckedState } from './base/option-checked-state';
+import { OptionInputs } from './base/option-inputs';
+import { OptionLabelSize } from './base/option-label-size';
+
+/**
+ * generates one or more html checkbox inputs
+ *
+ * @constructor
+ * @param {string} htmlId
+ * @param {string} htmlClass
+ * @param {string} value
+ * @param {Array} options
+ * @param {Function} onChange
+ * @param {boolean} btnGroup
+ * @param {string} helpTextID
+ * @param {Object} dataSet
+ * @param {Object} attributes
+ * @return {string} rendered date name form row
+ */
+export class CheckboxInput extends Component {
+	static propTypes = {
+		htmlId: PropTypes.string.isRequired,
+		htmlClass: PropTypes.string,
+		value: PropTypes.oneOfType( [
+			PropTypes.array,
+			PropTypes.number,
+			PropTypes.string,
+		] ),
+		options: PropTypes.array,
+		onChange: PropTypes.func,
+		btnGroup: PropTypes.bool,
+		helpTextID: PropTypes.string,
+		dataSet: PropTypes.object,
+	};
+
+	constructor( props ) {
+		super( props );
+		this.state = {
+			checked: this.resetState(),
+		};
+	}
+
+	/**
+	 * @function
+	 * @return {Object} state
+	 */
+	resetState = () => {
+		return OptionCheckedState(
+			this.props.htmlId,
+			this.props.options,
+			this.props.initialValue,
+		);
+	};
+
+	/**
+	 * @function
+	 * @param {Object} event
+	 */
+	toggleChecked = ( event ) => {
+		if ( event.target &&
+			event.target.id &&
+			event.target.checked !== undefined
+		) {
+			const checked = this.state.checked;
+			checked[ event.target.id ] = event.target.checked;
+			this.setState( { checked: checked } );
```

Comment:
`OptionInputs` is a functional component so refs won't work (as is) because refs can only be used with class components. Making `OptionInputs` a class component is an easy enough change, but what is the benefit of doing so? 

From the article that you linked to:

> Generally, the use of refs should be considered only when the required interaction cannot be achieved using the mechanisms of state and props.
> 
> However, there are a couple of cases where using a ref is appropriate. One of which is when integrating with third-party DOM libraries. Also, deep interactions such as handling text selections or managing media playback behavior also require the use of refs on the corresponding elements.

My understanding is that refs give you direct access to a component or DOM element and I don't see how that would benefit us here. Currently `this.state.checked` is an array where the keys are input IDs and the values are a boolean reflecting their checked state, which is really simple to manage using the `onClick` event.

Just so everything is clear, the above code is for handling state for the entire group of checkbox or radio inputs. As well, although it is impossible to tell from just this code, I am using css to generate really mobile friendly inputs ala Bootstrap button groups: https://getbootstrap.com/docs/3.3/components/#btn-groups-single
so the click handler is actually on the label element (which also triggers the checked state of the input itself).

If anything, I could see moving the state handling down to the individual inputs generated by `OptionInputs`. ie: refactor `OptionInputs` to be a singular `OptionInput` component (with state) and perform the options mapping in the `CheckboxInput` component, however that would require having to duplicate some of this logic in the `RadioInput` as well. The primary reason I did things the way I did was so that `CheckboxInput` and `RadioInput` could be as minimal as possible and let `OptionInputs` handle the options mapping, so making that change would potentially result in some code duplication.

I'm not trying to argue against using refs I just literally don't get how that could be done in a way that would benefit this component.

I would see the following changes being necessary:

- both `CheckboxInput` and `RadioInput` would have to maintain an array of refs to all of the inputs generated for their options. This could potentially be offloaded into another component similar to `OptionCheckedState` that both input types could share in order to avoid duplication.
- we would still need to also maintain an array for the checked state, unless your intent is to do away with that and use the refs to obtain those values directly (I'm guessing this is the main focus of your suggestion)
- regardless of the previous point, we would still also need a callback for handling click events, but instead of getting the checked state from the event, we would get it from the input itself... but then... what do we do with that? ya i just don't get how this would work or what the benefit would be...

Can you elaborate on your suggestion plz?

Reply:
In the interest of time, I don't have a practical example to offer you right now and you're right it would require refactoring how you are doing things here somewhat.  It's something we can iterate on later if it turns out its needed.